File: code/ex1025.java
// DFS + memo

class Solution {
    public boolean divisorGame(int N) {
        boolean[] memo = new boolean[N + 1];
        return dfs(N, memo);
    }
    
    private boolean dfs(int N, boolean[] memo) {    // Player: 1 => Alice; -1 => Bob
        if (N <= 1)    return false;
        if (memo[N])    return true;
        boolean res = false;
        for (int x = 1; x < N; x++) {
            if (N % x != 0)    continue;
            if (!dfs(N - x, memo)) {res = true;    break;}
        }
        memo[N] = res;
        return res;
    }
}

File: code/ex1025.cpp
class Solution {
public:
    bool divisorGame(int N) {
        vector<int> dp(N + 1, -1);
        return backtracking(N, dp);
    }
    
private:
    bool backtracking(int N, vector<int>& dp) {
        if (N <= 1)    return 0;
        if (dp[N] != -1)    return dp[N] == 1 ? 1 : 0;
        bool res = 0;
        for (int x = 1; x <= N / 2; x++) {
            if (N % x != 0)    continue;
            res |= !backtracking(N - x, dp);
            if (res)    break;
        }
        dp[N] = res;
        return res;
    }
};